src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,370 @@
+    size_t limitDescendantsCount = GetArg("-limitdescendantcount", BU_DEFAULT_DESCENDANT_LIMIT);
+    size_t limitDescendantSize = GetArg("-limitdescendantsize", BU_DEFAULT_DESCENDANT_SIZE_LIMIT) * 1000;
+    std::string errString;
+
+    READLOCK(mempool.cs_txmempool);
+    bool ret = mempool.ValidateMemPoolAncestors(
+        txIn, limitAncestorCount, limitAncestorSize, limitDescendantsCount, limitDescendantSize, errString);
+    if (!ret)
+    {
+        LOG(SELECTCOINS, "CoinSelection eliminated a solution, error: %s\n", errString.c_str());
+    }
+    return ret;
+
+    // return true;
+}
+
+// Select coins.
+TxoGroup CoinSelection(/*const*/ SpendableTxos &available,
+    const CAmount targetValue,
+    const CAmount &dust,
+    CFeeRate feeRate,
+    unsigned int changeLen)
+{
+    FastRandomContext insecureRand;
+    uint64_t utxosInWallet = available.size();
+    SpendableTxos::iterator aEnd = available.end();
+    SpendableTxos::iterator aBegin = available.begin();
+    TxoGroupMap solutions;
+    TxoGroupMap candidates;
+
+    LOG(SELECTCOINS, "CoinSelection: Target: %d, num available txos: %d\n", targetValue, available.size());
+
+    // Find the smallest output > TargetValue & add it to solutions (if it exists).
+    SpendableTxos::iterator large;
+    large = available.lower_bound(targetValue); // Find the elem nearest the target but greater or =
+    // The amount is bigger than our biggest output.  We'll create a simple solution from a set of our biggest outputs.
+    if (large == aEnd)
+    {
+        if (large == aBegin) // beginning == ending, there is nothing
+        {
+            return TxoGroup(0, TxoItVec());
+        }
+        --large;
+        SpendableTxos::iterator i = large;
+        CAmount total = large->first;
+        TxoGroup group = makeGroup(i);
+        while ((total < targetValue) && (i != aBegin))
+        {
+            --i;
+            total += i->first;
+            group = makeGroup(i, &group);
+        }
+        if (total >= targetValue)
+        {
+            solutions.insert(TxoGroupMap::value_type(total, group));
+        }
+        else // We added every transaction and it did not sum to totalValue, so no solution.
+        {
+            LOG(SELECTCOINS, "Every available UTXO sums to %llu which is lower than the target %llu\n", total,
+                targetValue);
+            return TxoGroup(0, TxoItVec());
+        }
+    }
+    else // otherwise, the next one could be a solution
+    {
+        SpendableTxos::iterator i = large;
+        if (i != aEnd)
+        {
+            assert(i->first >= targetValue);
+            if (i->first == targetValue) // If its equal then done
+            {
+                return makeGroup(i);
+            }
+            // If its greater then add it to the solutions list.
+            solutions.insert(TxoGroupMap::value_type(i->first, makeGroup(i)));
+            // bestValue = i->first;
+        }
```

Comment:
some commented out code  // bestValue could be removed.

Reply:
removed